Allow device to be shared among threads for macOS backend.

Tederean opened the pull request:
This patch ensures that the backend implements Sync, as is already the case with Linux and Windows.

This has not worked so far because ‘Box’ was saved in the backend, which must not be shared between threads in the secure Rust context. Since the data is only provided for the callback that is called from C, I decided to save only the raw pointer/data as usize and to free manually in Drop.

This also guarantees that nobody accesses the data in parallel to the callback during operation.

Tested successfully on a Mac Studio M1 Max.

sidit77 commented on src/backend/iohidmanager/device.rs:
```diff
@@ -117,38 +117,50 @@ impl IOHIDDevice {
 
-type InputReportCallback = Box<dyn FnMut(&[u8]) + Send>;
+type InputReportCallback = Box<dyn FnMut(&[u8]) + Send + Sync>;
+type InputReportCallbackContainer = Box<UnsafeCell<InputReportCallback>>;
```
I think I overthought things when I wrote this and we don't need the `UnsafeCell`, especially not the way I used it.

sidit77 commented on src/backend/iohidmanager/device.rs:
```diff
@@ -158,6 +170,10 @@ impl Drop for CallbackGuard {
         unsafe {
             IOHIDDeviceRegisterInputReportCallback(self.device.as_concrete_TypeRef(), null_mut(), 0, transmute(null::<()>()), null_mut())
```
We should also remove this, Casting null to fn is very bad and it was only suppost to be a quick hack until the function prototype is fixed (https://github.com/jtakakura/io-kit-rs/issues/6), but it seems like this won't happen in the near future, so I think we should just "embed" the correct protoype inline:
```rust
unsafe {
    extern "C" { fn IOHIDDeviceRegisterInputReportCallback(device: IOHIDDeviceRef, report: *mut u8, reportLength: CFIndex, callback: Option<IOHIDReportCallback>, context: *mut c_void); }
    IOHIDDeviceRegisterInputReportCallback(self.device.as_concrete_TypeRef(), null_mut(), 0, None, null_mut())
}
```

sidit77 reviewed (changes requested):
Thank you for looking at the MacOs backend. It's definitly the weakest part of this library in my opinion, as I don't own a Mac and had no experience developing for it either.

Most of the suggestions are things that were my fault, but I think we can just roll them into the PR to fix everything wrong with this section of code at once.

Tederean commented:
I have added your objections and suggestions. I don't own a Mac either, but I have a working cross-compiling setup on which the changes compile. A test is still pending, I have only occasional access to a Mac.

I have now stored the pointers itself in the CallbackGuard struct and implemented the unsafe Sync + Send trait manually for this reason. This saves us having to cast back and forth and increases type safety.

Tederean commented:
Tested successfully on macOS. I think we can merge from my point of view.
